Avoid using CS2 allocators during compilation

ashu-mehra opened the pull request:
CS2 allocators create unnecessary overhead during memory allocation.
CS2 allocators like heap_allocator cause hogging of memory, thus
increasing memory usage during compilation.
Since all the CS2 memory allocators are using TR::Region as the backing
allocator, it is better to use TR::Region directly.
Also cleaned up some other dead code after the removal of the
CS2 allocators.

Signed-off-by: Ashutosh Mehra <mehra.ashutosh@ibm.com>

ashu-mehra commented:
This PR is for fixing issue https://github.com/eclipse/openj9/issues/7543

fjeremic commented:
Requesting @dsouzai for his expertise in the area to review.

ashu-mehra commented on compiler/env/Region.cpp:
```diff
@@ -90,6 +90,15 @@ Region::allocate(size_t const size, void *hint)
+{
+   void * npointer = allocate(newsize);
+   memcpy(npointer, pointer, newsize < size ? newsize : size);
+   deallocate(pointer, size);
```
I agree, it would be better if the `Region` provides some way to free the memory and maintain the free list internally. I think the issue you have opened #1995 is ideal to carry out that work.

ashu-mehra commented:
> However, in all of these places, it's a template parameter and when I dug through, the uses of the type is a reference. 

That's an important point and I missed out raising it here. Thanks for mentioning it.

andrewcraik commented:
The CS2 allocator as I remember it always had a capability to free memory and to reuse that freed memory - is there a risk here of having a significant growth in footprint because of the change to the allocator? I think some measurements are necessary before this is merged to make sure that metrics are not going to be blown out of the water.

andrewcraik commented:
For places that are using the CS2 allocator for non-CS2 structures then a switch to regions, possibly with some more narrowly scoped regions makes sense. For the CS2 structures I'm less sure...

dsouzai commented:
> The CS2 allocator as I remember it always had a capability to free memory and to reuse that freed memory - is there a risk here of having a significant growth in footprint because of the change to the allocator?

Given that the CS2 allocator would allocate 64 K chunks for each size range, I wouldn't be surprised if the switch to Region (which doesn't deallocate) on average balanced out. Still, definitely worth doing footprint and compilation memory analysis since this is just me speculating.

ashu-mehra commented:
I did Daytrader7 startup runs for java 8 and java 11 (2 batches of 5 iterations each) with following options:
`-Xquickstart -Xmx256m -Xshareclasses:enableBCI,name=liberty -Xscmx60M -Xscmaxaot8m`
Here are the numbers:
```
Results for JDK=/home/ashutosh/java8_base/j2sdk-image jvmOpts=-Xquickstart -Xmx256m -Xshareclasses:enableBCI,name=liberty -Xscmx60M -Xscmaxaot8m 
StartupTime	avg=5845	min=5663	max=6301	stdDev=205.1	maxVar=11.3%	confInt=2.03%	samples= 10
Footprint	avg=226955	min=221020	max=234704	stdDev=4885.3	maxVar=6.2%	confInt=1.25%	samples= 10
CThreadTime	avg=1860	min=1655	max=2254	stdDev=168.1	maxVar=36.2%	confInt=5.24%	samples= 10
Results for JDK=/home/ashutosh/java8_new/j2sdk-image jvmOpts=-Xquickstart -Xmx256m -Xshareclasses:enableBCI,name=liberty -Xscmx60M -Xscmaxaot8m 
StartupTime	avg=5840	min=5639	max=6228	stdDev=212.3	maxVar=10.4%	confInt=2.11%	samples= 10
Footprint	avg=228554	min=221852	max=246540	stdDev=8472.9	maxVar=11.1%	confInt=2.15%	samples= 10
CThreadTime	avg=2113	min=1953	max=2340	stdDev=131.7	maxVar=19.8%	confInt=3.61%	samples= 10
Results for JDK=/home/ashutosh/java11_base/jdk jvmOpts=-Xquickstart -Xmx256m -Xshareclasses:enableBCI,name=liberty -Xscmx60M -Xscmaxaot8m 
StartupTime	avg=6165	min=5996	max=6451	stdDev=157.2	maxVar=7.6%	confInt=1.48%	samples= 10
Footprint	avg=233037	min=228112	max=243768	stdDev=5168.6	maxVar=6.9%	confInt=1.29%	samples= 10
CThreadTime	avg=2316	min=2023	max=2665	stdDev=169.8	maxVar=31.7%	confInt=4.25%	samples= 10
Results for JDK=/home/ashutosh/java11_new/jdk jvmOpts=-Xquickstart -Xmx256m -Xshareclasses:enableBCI,name=liberty -Xscmx60M -Xscmaxaot8m 
StartupTime	avg=6271	min=6075	max=6576	stdDev=188.3	maxVar=8.3%	confInt=1.74%	samples= 10
Footprint	avg=234902	min=225940	max=251624	stdDev=9411.2	maxVar=11.4%	confInt=2.32%	samples= 10
CThreadTime	avg=2449	min=2241	max=2630	stdDev=123.4	maxVar=17.4%	confInt=2.92%	samples= 10
```

0.7% increase in footprint for java 8 and 0.8% increase in footprint for java 11.
No impact on startup time in java8, but 1.7% increase in java 11. Not sure why there is this difference between java 8 and java 11 with respect to impact on startup time.

dsouzai commented:
```
confInt=2.03%
confInt=2.11%
confInt=1.48%
confInt=1.74%
```
I think you're gonna have to do way more runs, more like 40+ samples.

ashu-mehra commented:
how low should the confInt be for acceptable results? < 1%?

dsouzai commented:
Yeah, ideally <1%

andrewcraik commented:
Also some of those CS2 structures/allocations are not hit during startup so a measure of footprint on high opt compiles is also warranted so something like HiBench or similar would be needed to ensure that you aren't adversely affecting stuff outside of startup.

ashu-mehra commented:
Another set of results for DT7 with more iterations:

```
Results for JDK=/home/ashutosh/java8_base/j2sdk-image jvmOpts=-Xquickstart -Xmx256m -Xshareclasses:enableBCI,name=liberty -Xscmx60M -Xscmaxaot8m 
StartupTime	avg=5749	min=5491	max=6239	stdDev=179.1	maxVar=13.6%	confInt=0.65%	samples= 63
	Outlier values:  6424.225
Footprint	avg=225620	min=213456	max=243104	stdDev=6015.0	maxVar=13.9%	confInt=0.56%	samples= 64
CThreadTime	avg=1879	min=1570	max=2268	stdDev=157.0	maxVar=44.5%	confInt=1.74%	samples= 64
Results for JDK=/home/ashutosh/java8_new/j2sdk-image jvmOpts=-Xquickstart -Xmx256m -Xshareclasses:enableBCI,name=liberty -Xscmx60M -Xscmaxaot8m 
StartupTime	avg=5788	min=5562	max=6343	stdDev=163.1	maxVar=14.1%	confInt=0.59%	samples= 64
Footprint	avg=230697	min=213116	max=246852	stdDev=9169.7	maxVar=15.8%	confInt=0.83%	samples= 64
CThreadTime	avg=2037	min=1741	max=2406	stdDev=151.4	maxVar=38.2%	confInt=1.55%	samples= 64
Results for JDK=/home/ashutosh/java11_base/jdk jvmOpts=-Xquickstart -Xmx256m -Xshareclasses:enableBCI,name=liberty -Xscmx60M -Xscmaxaot8m 
StartupTime	avg=6104	min=5832	max=6576	stdDev=180.1	maxVar=12.7%	confInt=0.61%	samples= 64
Footprint	avg=230329	min=221916	max=240212	stdDev=4868.2	maxVar=8.2%	confInt=0.44%	samples= 64
CThreadTime	avg=2219	min=1949	max=2579	stdDev=139.0	maxVar=32.3%	confInt=1.30%	samples= 64
Results for JDK=/home/ashutosh/java11_new/jdk jvmOpts=-Xquickstart -Xmx256m -Xshareclasses:enableBCI,name=liberty -Xscmx60M -Xscmaxaot8m 
StartupTime	avg=6168	min=5890	max=6682	stdDev=176.0	maxVar=13.4%	confInt=0.59%	samples= 64
Footprint	avg=236190	min=223524	max=258176	stdDev=9708.5	maxVar=15.5%	confInt=0.86%	samples= 64
CThreadTime	avg=2342	min=2043	max=2822	stdDev=151.9	maxVar=38.1%	confInt=1.35%	samples= 64
```

Here the `confInt` for StartupTime and Footprint is below 1%, but not for CThreadTime.
Footprint shows increase of 2.25% for java 8 and 2.5% for java 11.
StartupTime increased by 0.7% for java 8 and 1% for java 11.

So it does seem to have non-negligible impact on memory footprint.

dsouzai commented:
> So it does seem to have non-negligible impact on memory footprint.

fyi @mpirvu @vijaysun-omr

mpirvu commented:
Because scratch memory is completely released after a compilation I find it unlikely that this change will increase footprint measured after start-up.
CS2 allocators were introduced during the days when the code was shared between the static and the dynamic compilers. The static compiler used a lot of memory and we needed the ability to reuse some of the memory that was freed during a compilation. This had the advantage of a lower memory peak but an increased in CPU because of the need to manage free blocks. My expectation from this change is a potential increase peak in scratch memory usege and an improvement in CPU consumption (possibly an improvement in start-up time as well)

andrewcraik commented:
I would worry that we might get more compilation aborts due to the increased peak memory usage - esp under high opt JSR292 compiles.

mpirvu commented:
@ashu-mehra is it possible to do this on a per-compilation basis? If so, it could be applied just to cold/noOpt compilations

dsouzai commented:
> is it possible to do this on a per-compilation basis?

I don't think so; the `TR::Allocator` is a C++ type definition that gets resolved at compile time. We'd have to do some pretty intense template magic, maybe.

andrewcraik commented:
Rather than replacing the allocator, would it make more sense to replace the CS2 data structures that are responsible for the memory consumption and replace them with TR/STL equivalents using the region allocator? That way the data structure and the allocator remain matched and the changes could be done in a more targeted way...

andrewcraik commented:
Could the commits or even the PRs for the be broken up to those that are adjusting the use of a Region by adding/removing &s etc and those that are related to the CS2 allocator conversion?

mpirvu commented:
DT7 start-up numbers with -Xquickstart (liberty-wp7-19.0.0.9) and with the same builds used by @ashu-mehra in https://github.com/eclipse/omr/pull/4612#issuecomment-559879929.
```
Results for JDK=/home/mpirvu/sdks/Ashutosh/baseline jvmOpts=-Xquickstart -Xmx256m
StartupTime     avg=3539        min=3298        max=3869        stdDev=108.0    maxVar=17.3%    confInt=0.41%   samples=154
Footprint       avg=205194      min=183208      max=233264      stdDev=11781.2  maxVar=27.3%    confInt=0.75%   samples=160
CThreadTime     avg= 780        min= 601        max=1156        stdDev=116.6    maxVar=92.3%    confInt=1.96%   samples=158
ProcessTime     avg=6008        min=5240        max=7210        stdDev=419.2    maxVar=37.6%    confInt=0.91%   samples=160
Stats for cold run:
StartupTime     avg=9292        min=9029        max=9557        stdDev=147.0    maxVar=5.9%     confInt=0.63%   samples= 19
Footprint       avg=216598      min=204908      max=247796      stdDev=12731.4  maxVar=20.9%    confInt=2.27%   samples= 20
CThreadTime     avg=6598        min=6276        max=6987        stdDev=215.5    maxVar=11.3%    confInt=1.26%   samples= 20
ProcessTime     avg=18998       min=18330       max=20010       stdDev=476.0    maxVar=9.2%     confInt=0.97%   samples= 20
Results for JDK=/home/mpirvu/sdks/Ashutosh/nocs2 jvmOpts=-Xquickstart -Xmx256m
StartupTime     avg=3560        min=3314        max=3895        stdDev=104.6    maxVar=17.6%    confInt=0.40%   samples=146
Footprint       avg=203601      min=179452      max=229816      stdDev=10396.7  maxVar=28.1%    confInt=0.67%   samples=160
CThreadTime     avg= 782        min= 624        max=1131        stdDev=90.1     maxVar=81.2%    confInt=1.51%   samples=158
ProcessTime     avg=6005        min=5270        max=7020        stdDev=350.4    maxVar=33.2%    confInt=0.76%   samples=159
Stats for cold run:
StartupTime     avg=9239        min=8915        max=9552        stdDev=146.9    maxVar=7.1%     confInt=0.63%   samples= 19
Footprint       avg=217365      min=202328      max=240096      stdDev=11615.7  maxVar=18.7%    confInt=2.07%   samples= 20
CThreadTime     avg=6448        min=5944        max=6947        stdDev=311.2    maxVar=16.9%    confInt=1.87%   samples= 20
ProcessTime     avg=18818       min=17990       max=19710       stdDev=503.7    maxVar=9.6%     confInt=1.03%   samples= 20
```
Summary for the warm runs:  0.59% larger start-up time, 0.8% smaller footprint, no change in CPU. Considering the confidence interval all these are noise.

ashu-mehra commented:
Did a single run of HiBench using Rahil's setup that runs seven different benchmarks to get numbers on compilation failures.
Across all the seven benchmarks, the baseline build reported 5 compilation failures due to `compilationHeapLimitExceeded`.
With this PR there are 135 compilation failures due to `compilationHeapLimitExceeded`.

These are the failures with baseline build (all failures clubbed together):
```
! org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.write(Lscala/collection/Iterator;)V time=881476us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228578 MB mem=[region=262144 system=262144]KB
! org/apache/spark/graphx/impl/EdgePartition.updateVertices(Lscala/collection/Iterator;)Lorg/apache/spark/graphx/impl/EdgePartition; time=1503279us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=219625 MB mem=[region=262144 system=262144]KB
! org/apache/spark/storage/memory/MemoryStore.putIteratorAsValues(Lorg/apache/spark/storage/BlockId;Lscala/collection/Iterator;Lscala/reflect/ClassTag;)Lscala/util/Either; time=1550781us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=227009 MB mem=[region=262144 system=262144]KB
! com/twitter/chill/ScalaCollectionsRegistrar.apply(Lcom/esotericsoftware/kryo/Kryo;)V time=1447014us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229967 MB mem=[region=262144 system=262144]KB
! org/apache/spark/examples/mllib/SparseNaiveBayes$$anonfun$15$$anonfun$18.apply(Ljava/lang/Object;)Ljava/lang/Object; time=615419us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229914 MB mem=[region=262144 system=262144]KB
```

These are the failures with this PR (all failures clubbed together):
```
! com/intel/hibench/sparkbench/graph/nweight/GraphxNWeight$$anonfun$1$$anonfun$apply$2.apply(Ljava/lang/String;)Lscala/Tuple2; time=284577us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229900 MB mem=[region=259072 system=262144]KB
! com/esotericsoftware/kryo/Kryo.writeClassAndObject(Lcom/esotericsoftware/kryo/io/Output;Ljava/lang/Object;)V time=1031560us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229805 MB mem=[region=217088 system=229376]KB
! java/lang/String.split(Ljava/lang/String;I)[Ljava/lang/String; time=564204us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229805 MB mem=[region=256128 system=262144]KB
! java/util/regex/Pattern.split(Ljava/lang/CharSequence;I)[Ljava/lang/String; time=257760us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229916 MB mem=[region=258880 system=262144]KB
! com/intel/hibench/sparkbench/graph/nweight/GraphxNWeight$$anonfun$1.apply(Ljava/lang/String;)Lscala/collection/mutable/ArrayOps; time=327790us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229957 MB mem=[region=252544 system=262144]KB
! com/esotericsoftware/kryo/util/DefaultClassResolver.writeName(Lcom/esotericsoftware/kryo/io/Output;Ljava/lang/Class;Lcom/esotericsoftware/kryo/Registration;)V time=361150us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229941 MB mem=[region=259264 system=262144]KB
! java/util/regex/Pattern.compile()V time=625731us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229883 MB mem=[region=258368 system=262144]KB
! org/apache/hadoop/io/Text.decode(Ljava/nio/ByteBuffer;Z)Ljava/lang/String; time=432274us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229810 MB mem=[region=261824 system=262144]KB
! org/apache/hadoop/util/LineReader.readDefaultLine(Lorg/apache/hadoop/io/Text;II)I time=407087us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229942 MB mem=[region=259136 system=262144]KB
! scala/collection/Iterator$$anon$12.nextCur()V time=304188us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229932 MB mem=[region=256960 system=262144]KB
! java/util/regex/Pattern.clazz(Z)Ljava/util/regex/Pattern$CharProperty; time=1220636us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229983 MB mem=[region=255552 system=262144]KB
! org/apache/hadoop/io/Text.decode([BII)Ljava/lang/String; time=288316us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229909 MB mem=[region=257856 system=262144]KB
! sun/reflect/MethodAccessorGenerator.generate(Ljava/lang/Class;Ljava/lang/String;[Ljava/lang/Class;Ljava/lang/Class;[Ljava/lang/Class;IZZLjava/lang/Class;)Lsun/reflect/MagicAccessorImpl; time=546547us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229895 MB mem=[region=262144 system=262144]KB
! com/esotericsoftware/kryo/serializers/FieldSerializer.read(Lcom/esotericsoftware/kryo/Kryo;Lcom/esotericsoftware/kryo/io/Input;Ljava/lang/Class;)Ljava/lang/Object; time=244852us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229813 MB mem=[region=257472 system=262144]KB
! com/esotericsoftware/kryo/Kryo.readClassAndObject(Lcom/esotericsoftware/kryo/io/Input;)Ljava/lang/Object; time=306160us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229879 MB mem=[region=258944 system=262144]KB
! scala/collection/Iterator$$anon$11.next()Ljava/lang/Object; time=197831us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229750 MB mem=[region=257088 system=262144]KB
! com/esotericsoftware/kryo/util/DefaultClassResolver.readName(Lcom/esotericsoftware/kryo/io/Input;)Lcom/esotericsoftware/kryo/Registration; time=758006us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229839 MB mem=[region=241664 system=245760]KB
! org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.write(Lscala/collection/Iterator;)V time=793561us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229817 MB mem=[region=219904 system=229376]KB
! com/esotericsoftware/kryo/util/DefaultClassResolver.writeClass(Lcom/esotericsoftware/kryo/io/Output;Ljava/lang/Class;)Lcom/esotericsoftware/kryo/Registration; time=581951us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229822 MB mem=[region=213568 system=229376]KB
! org/apache/spark/util/CompletionIterator.next()Ljava/lang/Object; time=191275us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229934 MB mem=[region=256832 system=262144]KB
! org/apache/spark/util/collection/TimSort.binarySort(Ljava/lang/Object;IIILjava/util/Comparator;)V time=389295us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229436 MB mem=[region=259520 system=262144]KB
! org/apache/spark/util/collection/TimSort$SortState.gallopLeft(Ljava/lang/Object;Ljava/lang/Object;IIILjava/util/Comparator;)I time=377887us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229385 MB mem=[region=262080 system=262144]KB
! org/apache/spark/util/collection/TimSort$SortState.gallopRight(Ljava/lang/Object;Ljava/lang/Object;IIILjava/util/Comparator;)I time=428024us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229592 MB mem=[region=260544 system=262144]KB
! org/apache/spark/util/collection/TimSort$SortState.mergeHi(IIII)V time=635956us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229797 MB mem=[region=254848 system=262144]KB
! org/apache/spark/util/collection/AppendOnlyMap.changeValue(Ljava/lang/Object;Lscala/Function2;)Ljava/lang/Object; time=255212us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229604 MB mem=[region=258816 system=262144]KB
! org/apache/spark/util/collection/TimSort$SortState.mergeLo(IIII)V time=472476us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229604 MB mem=[region=260608 system=262144]KB
! org/apache/spark/util/collection/OpenHashSet.addWithoutResize(Ljava/lang/Object;)I time=300229us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228862 MB mem=[region=258112 system=262144]KB
! org/apache/spark/util/collection/OpenHashSet.rehash(Ljava/lang/Object;Lscala/Function1;Lscala/Function2;)V time=343121us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228785 MB mem=[region=259648 system=262144]KB
! com/esotericsoftware/kryo/serializers/DefaultArraySerializers$ObjectArraySerializer.write(Lcom/esotericsoftware/kryo/Kryo;Lcom/esotericsoftware/kryo/io/Output;[Ljava/lang/Object;)V time=1251904us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228740 MB mem=[region=198016 system=212992]KB
! scala/collection/Iterator$class.foreach(Lscala/collection/Iterator;Lscala/Function1;)V time=1184315us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228842 MB mem=[region=179904 system=180224]KB
! org/apache/spark/graphx/impl/RoutingTablePartition$$anonfun$fromMsgs$1.apply(Ljava/lang/Object;)Ljava/lang/Object; time=234281us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228407 MB mem=[region=259008 system=262144]KB
! org/apache/spark/serializer/DeserializationStream$$anon$2.getNext()Ljava/lang/Object; time=185092us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228126 MB mem=[region=254464 system=262144]KB
! com/esotericsoftware/kryo/serializers/DefaultArraySerializers$ObjectArraySerializer.read(Lcom/esotericsoftware/kryo/Kryo;Lcom/esotericsoftware/kryo/io/Input;Ljava/lang/Class;)Ljava/lang/Object; time=265271us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228081 MB mem=[region=259136 system=262144]KB
! com/esotericsoftware/kryo/Kryo.readObject(Lcom/esotericsoftware/kryo/io/Input;Ljava/lang/Class;Lcom/esotericsoftware/kryo/Serializer;)Ljava/lang/Object; time=343715us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=227759 MB mem=[region=257408 system=262144]KB
! org/apache/spark/graphx/util/collection/GraphXPrimitiveKeyOpenHashMap.changeValue(Ljava/lang/Object;Lscala/Function0;Lscala/Function1;)Ljava/lang/Object; time=287002us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=227767 MB mem=[region=259520 system=262144]KB
! org/apache/spark/graphx/impl/ShippableVertexPartition$$anonfun$shipVertexAttributes$1$$anonfun$apply$1.apply$mcVJ$sp(J)V time=301377us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=227775 MB mem=[region=255936 system=262144]KB
! com/esotericsoftware/kryo/util/IdentityObjectIntMap.push(Ljava/lang/Object;IILjava/lang/Object;ILjava/lang/Object;ILjava/lang/Object;)V time=444564us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=227583 MB mem=[region=261248 system=262144]KB
! scala/collection/Iterator$$anon$12.next()Ljava/lang/Object; time=1606463us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=227817 MB mem=[region=198848 system=212992]KB
! com/esotericsoftware/kryo/Kryo.writeObject(Lcom/esotericsoftware/kryo/io/Output;Ljava/lang/Object;Lcom/esotericsoftware/kryo/Serializer;)V time=356920us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=227761 MB mem=[region=256192 system=262144]KB
! com/esotericsoftware/kryo/Kryo.writeReferenceOrNull(Lcom/esotericsoftware/kryo/io/Output;Ljava/lang/Object;Z)Z time=249660us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=227765 MB mem=[region=260480 system=262144]KB
! org/apache/spark/util/collection/TimSort.sort(Ljava/lang/Object;IILjava/util/Comparator;)V time=295264us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=225867 MB mem=[region=252800 system=262144]KB
! org/apache/spark/graphx/impl/EdgePartitionBuilder.toEdgePartition()Lorg/apache/spark/graphx/impl/EdgePartition; time=372160us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=225850 MB mem=[region=242112 system=245760]KB
! org/apache/spark/util/collection/OpenHashSet.getPos(Ljava/lang/Object;)I time=277771us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=224467 MB mem=[region=259904 system=262144]KB
! org/apache/spark/graphx/impl/EdgePartition.updateVertices(Lscala/collection/Iterator;)Lorg/apache/spark/graphx/impl/EdgePartition; time=1075647us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=218734 MB mem=[region=206208 system=212992]KB
! org/apache/spark/graphx/impl/EdgePartition.aggregateMessagesEdgeScan(Lscala/Function1;Lscala/Function2;Lorg/apache/spark/graphx/TripletFields;Lorg/apache/spark/graphx/impl/EdgeActiveness;Lscala/reflect/ClassTag;)Lscala/collection/Iterator; time=421083us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=216975 MB mem=[region=260672 system=262144]KB
! com/intel/hibench/sparkbench/graph/nweight/GraphxNWeight$.mapF(Lorg/apache/spark/graphx/EdgeContext;)V time=1295726us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=217177 MB mem=[region=196480 system=196608]KB
! scala/collection/immutable/Range.foreach(Lscala/Function1;)V time=439034us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=216556 MB mem=[region=259136 system=262144]KB
! com/esotericsoftware/kryo/serializers/MapSerializer.write(Lcom/esotericsoftware/kryo/Kryo;Lcom/esotericsoftware/kryo/io/Output;Ljava/lang/Object;)V time=221610us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=213355 MB mem=[region=256640 system=262144]KB
! com/esotericsoftware/kryo/serializers/MapSerializer.read(Lcom/esotericsoftware/kryo/Kryo;Lcom/esotericsoftware/kryo/io/Input;Ljava/lang/Class;)Ljava/lang/Object; time=232998us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=213294 MB mem=[region=256576 system=262144]KB
! com/esotericsoftware/kryo/util/DefaultClassResolver.readClass(Lcom/esotericsoftware/kryo/io/Input;)Lcom/esotericsoftware/kryo/Registration; time=526188us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=213274 MB mem=[region=258112 system=262144]KB
! com/intel/hibench/sparkbench/graph/nweight/GraphxNWeight$.updateF(JLcom/intel/hibench/sparkbench/graph/nweight/SizedPriorityQueue;Lscala/Option;)Lcom/intel/hibench/sparkbench/graph/nweight/SizedPriorityQueue; time=1082205us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=213138 MB mem=[region=181248 system=196608]KB
! org/apache/spark/util/collection/OpenHashSet.getPos$mcJ$sp(J)I time=224403us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=213050 MB mem=[region=261760 system=262144]KB
! com/esotericsoftware/kryo/Kryo.readClass(Lcom/esotericsoftware/kryo/io/Input;)Lcom/esotericsoftware/kryo/Registration; time=406405us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=213016 MB mem=[region=260608 system=262144]KB
! org/apache/spark/graphx/impl/VertexPartitionBaseOps.leftJoin(Lorg/apache/spark/graphx/impl/VertexPartitionBase;Lscala/Function3;Lscala/reflect/ClassTag;Lscala/reflect/ClassTag;)Lorg/apache/spark/graphx/impl/VertexPartitionBase; time=556149us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=213066 MB mem=[region=207872 system=212992]KB
! com/intel/hibench/sparkbench/graph/nweight/GraphxNWeight$$anonfun$nweight$2$$anonfun$apply$4.apply(Ljava/lang/Object;)Ljava/lang/Object; time=228528us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=212610 MB mem=[region=261568 system=262144]KB
! sun/misc/FloatingDecimal$BinaryToASCIIBuffer.dtoa(IJIZ)V time=593867us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=212707 MB mem=[region=261056 system=262144]KB
! org/apache/hadoop/io/Text.encode(Ljava/lang/String;Z)Ljava/nio/ByteBuffer; time=321189us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=212846 MB mem=[region=259584 system=262144]KB
! org/apache/mahout/math/VectorWritable.readFields(Ljava/io/DataInput;)V time=373602us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=232805 MB mem=[region=252992 system=262144]KB
! org/apache/hadoop/io/SequenceFile$Reader.next(Lorg/apache/hadoop/io/DataOutputBuffer;)I time=313662us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229666 MB mem=[region=259648 system=262144]KB
! com/intel/hibench/sparkbench/ml/DenseKMeans$$anonfun$7.apply(Lscala/Tuple2;)Lorg/apache/spark/mllib/linalg/Vector; time=406298us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229744 MB mem=[region=258368 system=262144]KB
! org/apache/mahout/math/Varint.readUnsignedVarInt(Ljava/io/DataInput;)I time=238636us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229717 MB mem=[region=257664 system=262144]KB
! org/apache/hadoop/hdfs/DFSInputStream.readWithStrategy(Lorg/apache/hadoop/hdfs/DFSInputStream$ReaderStrategy;II)I time=907919us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229760 MB mem=[region=257344 system=262144]KB
! org/apache/spark/rdd/HadoopRDD$$anon$1.getNext()Lscala/Tuple2; time=431086us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229534 MB mem=[region=257984 system=262144]KB
! org/apache/hadoop/hdfs/RemoteBlockReader2.read([BII)I time=409320us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228744 MB mem=[region=256320 system=262144]KB
! org/apache/hadoop/hdfs/DFSInputStream.read()I time=214525us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228309 MB mem=[region=257280 system=262144]KB
! com/intel/hibench/sparkbench/ml/DenseKMeans$$anonfun$7.apply(Ljava/lang/Object;)Ljava/lang/Object; time=320172us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=227069 MB mem=[region=258048 system=262144]KB
! org/apache/spark/util/collection/SizeTrackingVector.$plus$eq(Ljava/lang/Object;)V time=442485us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=226532 MB mem=[region=260672 system=262144]KB
! org/apache/spark/storage/BlockManager.getLocalValues(Lorg/apache/spark/storage/BlockId;)Lscala/Option; time=635361us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=226725 MB mem=[region=257856 system=262144]KB
! org/apache/spark/mllib/linalg/Vectors$.norm(Lorg/apache/spark/mllib/linalg/Vector;D)D time=413719us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=226781 MB mem=[region=260800 system=262144]KB
! org/apache/spark/storage/memory/MemoryStore.putIteratorAsValues(Lorg/apache/spark/storage/BlockId;Lscala/collection/Iterator;Lscala/reflect/ClassTag;)Lscala/util/Either; time=1816595us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=226920 MB mem=[region=262144 system=262144]KB
! org/apache/spark/util/Utils$.getIteratorSize(Lscala/collection/Iterator;)J time=273734us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=226733 MB mem=[region=260032 system=262144]KB
! org/apache/spark/rdd/RDD$$anonfun$zip$1$$anonfun$apply$27$$anon$2.hasNext()Z time=443019us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=226327 MB mem=[region=260864 system=262144]KB
! org/apache/spark/mllib/clustering/KMeans$$anonfun$initKMeansParallel$2.apply(Ljava/lang/Object;)Ljava/lang/Object; time=205987us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=226400 MB mem=[region=254848 system=262144]KB
! scala/collection/Iterator$class.foreach(Lscala/collection/Iterator;Lscala/Function1;)V time=330044us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=225862 MB mem=[region=261248 system=262144]KB
! org/apache/spark/mllib/clustering/KMeans$$anonfun$13.apply(Ljava/lang/Object;)Ljava/lang/Object; time=246902us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=225769 MB mem=[region=254592 system=262144]KB
! scala/collection/mutable/ArrayBuffer.foreach(Lscala/Function1;)V time=260044us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=225734 MB mem=[region=260800 system=262144]KB
! org/apache/spark/util/collection/AppendOnlyMap.changeValue(Ljava/lang/Object;Lscala/Function2;)Ljava/lang/Object; time=342370us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=225735 MB mem=[region=257024 system=262144]KB
! org/apache/spark/mllib/clustering/KMeans$$anonfun$6$$anonfun$apply$2.apply(Ljava/lang/Object;)Ljava/lang/Object; time=408059us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=225669 MB mem=[region=260544 system=262144]KB
! org/apache/spark/mllib/clustering/KMeansModel$$anonfun$computeCost$1.apply(Ljava/lang/Object;)Ljava/lang/Object; time=213753us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=225595 MB mem=[region=254784 system=262144]KB
! org/apache/spark/util/collection/AppendOnlyMap.changeValue(Ljava/lang/Object;Lscala/Function2;)Ljava/lang/Object; time=304645us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=234473 MB mem=[region=260480 system=262144]KB
! java/lang/String.split(Ljava/lang/String;I)[Ljava/lang/String; time=461749us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=230029 MB mem=[region=262144 system=262144]KB
! com/esotericsoftware/kryo/Kryo.readClassAndObject(Lcom/esotericsoftware/kryo/io/Input;)Ljava/lang/Object; time=386076us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=230037 MB mem=[region=258688 system=262144]KB
! scala/collection/immutable/HashMap$HashTrieMap.updated0(Ljava/lang/Object;IILjava/lang/Object;Lscala/Tuple2;Lscala/collection/immutable/HashMap$Merger;)Lscala/collection/immutable/HashMap; time=326693us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229806 MB mem=[region=260288 system=262144]KB
! org/apache/spark/examples/mllib/SparseNaiveBayes$$anonfun$15$$anonfun$18.apply(Ljava/lang/Object;)Ljava/lang/Object; time=552449us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229660 MB mem=[region=262144 system=262144]KB
! scala/collection/TraversableLike$$anonfun$groupBy$1.apply(Ljava/lang/Object;)Ljava/lang/Object; time=377184us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229908 MB mem=[region=258688 system=262144]KB
! scala/collection/TraversableLike$class.map(Lscala/collection/TraversableLike;Lscala/Function1;Lscala/collection/generic/CanBuildFrom;)Ljava/lang/Object; time=277548us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229925 MB mem=[region=258752 system=262144]KB
! scala/collection/mutable/HashTable$class.resize(Lscala/collection/mutable/HashTable;I)V time=364972us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229830 MB mem=[region=261760 system=262144]KB
! scala/collection/mutable/HashTable$class.findOrAddEntry(Lscala/collection/mutable/HashTable;Ljava/lang/Object;Ljava/lang/Object;)Lscala/collection/mutable/HashEntry; time=343564us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229845 MB mem=[region=261504 system=262144]KB
! scala/collection/immutable/HashMap$HashTrieMap.foreach(Lscala/Function1;)V time=215942us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229856 MB mem=[region=262144 system=262144]KB
! scala/collection/TraversableLike$class.groupBy(Lscala/collection/TraversableLike;Lscala/Function1;)Lscala/collection/immutable/Map; time=279994us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229727 MB mem=[region=256128 system=262144]KB
! scala/collection/IndexedSeqOptimized$class.foreach(Lscala/collection/IndexedSeqOptimized;Lscala/Function1;)V time=217568us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229157 MB mem=[region=260544 system=262144]KB
! java/net/URLClassLoader.defineClass(Ljava/lang/String;Lsun/misc/Resource;)Ljava/lang/Class; time=426016us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=236626 MB mem=[region=153728 system=163840]KB
! org/apache/spark/util/collection/AppendOnlyMap.changeValue(Ljava/lang/Object;Lscala/Function2;)Ljava/lang/Object; time=366338us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=231894 MB mem=[region=260224 system=262144]KB
! org/apache/spark/util/collection/AppendOnlyMap.growTable()V time=363750us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=231178 MB mem=[region=260480 system=262144]KB
! java/util/regex/Pattern.split(Ljava/lang/CharSequence;I)[Ljava/lang/String; time=327028us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=230870 MB mem=[region=254144 system=262144]KB
! scala/collection/IndexedSeqOptimized$class.slice(Lscala/collection/IndexedSeqOptimized;II)Ljava/lang/Object; time=435641us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229638 MB mem=[region=261952 system=262144]KB
! java/util/regex/Pattern.compile()V time=641212us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229786 MB mem=[region=258880 system=262144]KB
! org/apache/hadoop/io/Text.decode(Ljava/nio/ByteBuffer;Z)Ljava/lang/String; time=698783us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229534 MB mem=[region=261824 system=262144]KB
! java/lang/String.split(Ljava/lang/String;I)[Ljava/lang/String; time=777533us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229121 MB mem=[region=256128 system=262144]KB
! scala/Tuple2.equals(Ljava/lang/Object;)Z time=438370us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229136 MB mem=[region=261568 system=262144]KB
! org/apache/hadoop/util/LineReader.readDefaultLine(Lorg/apache/hadoop/io/Text;II)I time=282127us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229132 MB mem=[region=261824 system=262144]KB
! org/apache/spark/util/collection/ExternalSorter.getPartition(Ljava/lang/Object;)I time=256204us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228042 MB mem=[region=258688 system=262144]KB
! org/apache/spark/util/collection/TimSort.binarySort(Ljava/lang/Object;IIILjava/util/Comparator;)V time=266822us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=227784 MB mem=[region=260288 system=262144]KB
! org/apache/spark/util/collection/TimSort$SortState.mergeAt(I)V time=458506us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=227845 MB mem=[region=259264 system=262144]KB
! org/apache/spark/util/collection/TimSort$SortState.mergeHi(IIII)V time=339771us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=227978 MB mem=[region=259072 system=262144]KB
! org/apache/spark/util/collection/ExternalSorter$$anonfun$4$$anon$6.compare(Ljava/lang/Object;Ljava/lang/Object;)I time=327979us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228054 MB mem=[region=261376 system=262144]KB
! org/apache/spark/util/collection/TimSort.sort(Ljava/lang/Object;IILjava/util/Comparator;)V time=330469us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228047 MB mem=[region=262080 system=262144]KB
! com/esotericsoftware/kryo/Kryo.writeClassAndObject(Lcom/esotericsoftware/kryo/io/Output;Ljava/lang/Object;)V time=360862us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228035 MB mem=[region=255808 system=262144]KB
! com/esotericsoftware/kryo/Kryo.readClassAndObject(Lcom/esotericsoftware/kryo/io/Input;)Ljava/lang/Object; time=384259us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=228011 MB mem=[region=258112 system=262144]KB
! org/apache/spark/util/collection/ExternalAppendOnlyMap.insertAll(Lscala/collection/Iterator;)V time=277424us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=226580 MB mem=[region=261248 system=262144]KB
! org/apache/spark/util/SizeEstimator$.visitSingleObject(Ljava/lang/Object;Lorg/apache/spark/util/SizeEstimator$SearchState;)V time=330939us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=224830 MB mem=[region=256704 system=262144]KB
! org/apache/spark/util/SizeEstimator$.estimate(Ljava/lang/Object;Ljava/util/IdentityHashMap;)J time=568723us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=224951 MB mem=[region=262144 system=262144]KB
! org/apache/spark/util/SizeEstimator$SearchState.enqueue(Ljava/lang/Object;)V time=341456us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=224989 MB mem=[region=259392 system=262144]KB
! org/apache/spark/util/collection/SizeTrackingAppendOnlyMap.changeValue(Ljava/lang/Object;Lscala/Function2;)Ljava/lang/Object; time=219945us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=224541 MB mem=[region=255616 system=262144]KB
! org/apache/spark/util/collection/ExternalSorter.writePartitionedFile(Lorg/apache/spark/storage/BlockId;Ljava/io/File;)[J time=249165us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=224493 MB mem=[region=261056 system=262144]KB
! org/apache/spark/util/collection/ExternalSorter.insertAll(Lscala/collection/Iterator;)V time=797594us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=224206 MB mem=[region=223872 system=229376]KB
! java/net/URLClassLoader.defineClass(Ljava/lang/String;Lsun/misc/Resource;)Ljava/lang/Class; time=416876us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=236706 MB mem=[region=158720 system=163840]KB
! org/apache/spark/BaseRangePartitioner.getPartition(Ljava/lang/Object;)I time=335878us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=231817 MB mem=[region=259264 system=262144]KB
! com/esotericsoftware/kryo/Kryo.writeClassAndObject(Lcom/esotericsoftware/kryo/io/Output;Ljava/lang/Object;)V time=419833us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=230250 MB mem=[region=261120 system=262144]KB
! org/apache/hadoop/hdfs/DFSInputStream.read([BII)I time=372481us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229958 MB mem=[region=255296 system=262144]KB
! org/apache/hadoop/examples/terasort/TeraInputFormat$TeraRecordReader.nextKeyValue()Z time=367422us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229954 MB mem=[region=261312 system=262144]KB
! org/apache/spark/rdd/NewHadoopRDD$$anon$1.next()Lscala/Tuple2; time=451050us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=230083 MB mem=[region=229568 system=245760]KB
! com/esotericsoftware/kryo/Kryo.readClassAndObject(Lcom/esotericsoftware/kryo/io/Input;)Ljava/lang/Object; time=428968us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=229969 MB mem=[region=261760 system=262144]KB
! org/apache/spark/util/NextIterator.hasNext()Z time=172591us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=230039 MB mem=[region=257664 system=262144]KB
! org/apache/spark/util/collection/TimSort$SortState.mergeLo(IIII)V time=512626us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=226947 MB mem=[region=257984 system=262144]KB
! org/apache/spark/util/collection/TimSort.binarySort(Ljava/lang/Object;IIILjava/util/Comparator;)V time=293433us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=226803 MB mem=[region=260416 system=262144]KB
! scala/math/Ordering$$anon$9.compare(Ljava/lang/Object;Ljava/lang/Object;)I time=517715us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=226935 MB mem=[region=259968 system=262144]KB
! org/apache/spark/util/collection/TimSort$SortState.mergeHi(IIII)V time=520707us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=226953 MB mem=[region=258752 system=262144]KB
! org/apache/spark/rdd/PairRDDFunctions$$anonfun$saveAsNewAPIHadoopDataset$1$$anonfun$12$$anonfun$apply$4.apply$mcV$sp()V time=928147us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=225216 MB mem=[region=243840 system=245760]KB
! org/apache/hadoop/fs/FSDataOutputStream$PositionCache.write([BII)V time=298218us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=225230 MB mem=[region=255168 system=262144]KB
! org/apache/hadoop/hdfs/DFSOutputStream.writeChunkImpl([BII[BII)V time=436525us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=225205 MB mem=[region=259840 system=262144]KB
! java/net/URLClassLoader.defineClass(Ljava/lang/String;Lsun/misc/Resource;)Ljava/lang/Class; time=459008us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=236802 MB mem=[region=158720 system=163840]KB
! java/lang/String.split(Ljava/lang/String;I)[Ljava/lang/String; time=518164us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=231981 MB mem=[region=256448 system=262144]KB
! org/apache/spark/util/collection/AppendOnlyMap.changeValue(Ljava/lang/Object;Lscala/Function2;)Ljava/lang/Object; time=330346us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=231957 MB mem=[region=258688 system=262144]KB
! org/apache/hadoop/io/Text.decode(Ljava/nio/ByteBuffer;Z)Ljava/lang/String; time=294699us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=231630 MB mem=[region=259584 system=262144]KB
```

mpirvu commented:
There is something we are missing here because some of the compilations fail before reaching the 256M threshold. Example:
! java/net/URLClassLoader.defineClass(Ljava/lang/String;Lsun/misc/Resource;)Ljava/lang/Class; time=459008us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=236802 MB mem=[region=158720 system=163840]KB

andrewcraik commented:
At risk of stating the obvious - that kind of increase in failures is going to cause performance problems. In case it was missed - what about just tackling the CS2 data structures that are consuming the most memory and converting them to other alternatives rather than trying to change the allocator behind CS2?

mpirvu commented:
> that kind of increase in failures is going to cause performance problems

Indeed, that's a showstopper.

> tackling the CS2 data structures that are consuming the most memory

As far as I understand the problem lies in the implementation of the CS2 allocators. Small allocations of same size go into the same segment. So, if I need to allocate a few integers, a full segment is allocated. I need to allocate a few longs, another new segment is allocated and so on. This could be quite wasteful when dealing with very cheap compilations which is exactly what this PR tries to adress (https://github.com/eclipse/openj9/issues/7543)

andrewcraik commented:
> This could be quite wasteful when dealing with very cheap compilations which is exactly what this PR tries to adress 

Right but rather than trying to change the allocator under CS2 data structures convert the data structures to a more modern alternative (TR_ or STL) for the places that have a problem rather than changing the allocator without changing the data structure...

ashu-mehra commented:
> There is something we are missing here because some of the compilations fail before reaching the 256M threshold. 

I instrumented the code to print the size of the failed request:
```
diff --git a/runtime/compiler/env/SystemSegmentProvider.cpp b/runtime/compiler/env/SystemSegmentProvider.cpp
index 1e19bae..b40ff7d 100644
--- a/runtime/compiler/env/SystemSegmentProvider.cpp
+++ b/runtime/compiler/env/SystemSegmentProvider.cpp
@@ -99,6 +99,13 @@ J9::SystemSegmentProvider::request(size_t requiredSize)
    size_t systemSegmentSize = std::max(roundedSize, _systemSegmentSize);
    if (_systemBytesAllocated + systemSegmentSize > _allocationLimit )
       {
+      TR::Compilation *compiler = TR::comp();
+      if (TR::Options::isAnyVerboseOptionSet(TR_VerbosePerformance))
+         {
+         TR_VerboseLog::vlogAcquire();
+         TR_VerboseLog::writeLine(TR_Vlog_COMPFAIL,"%s _systemBytesAllocated=%zu, failedRequest=%zu, _allocationLimit=%zu", compiler->signature(), _systemBytesAllocated, systemSegmentSize, _allocationLimit
+         TR_VerboseLog::vlogRelease();
+         }
       throw std::bad_alloc();
       }
```

I wanted to verify that these failures are due to a very large request which would breach the `_allocationLimit` of 256M. Strangely the value `_systemBytesAllocated` displayed by above log message does not match with the `system` memory value displayed in the compilation log message, unless I misread the code.

Eg:
```
! org/apache/spark/graphx/impl/EdgePartition.updateVertices(Lscala/collection/Iterator;)Lorg/apache/spark/graphx/impl/EdgePartition; _systemBytesAllocated=259194880, failedRequest=24313856, _allocationLimit=268435456
! org/apache/spark/graphx/impl/EdgePartition.updateVertices(Lscala/collection/Iterator;)Lorg/apache/spark/graphx/impl/EdgePartition; time=708936us compilationHeapLimitExceeded memLimit=262144 KB freePhysicalMemory=214202 MB mem=[region=217856 system=229376]KB
```

Here `_systemBytesAllocated` is `259194880` which comes to around `253120 KB` but the next log message say `system=229376 KB`. Not sure how is that possible.
@mpirvu @dsouzai any idea?

ashu-mehra commented:
Probably we release some of the memory segments by the time we print the post compilation log message?

mpirvu commented:
> Probably we release some of the memory segments by the time we print the post compilation log message?

I was under the impression that we don't fully release memory segment except at the very end, but I could be wrong.

mpirvu commented:
Yet another infra failure:
```
07:48:05  Starting building: Pipeline_Build_Test_JDK8_ppc64le_linux #1326
Creating placeholder flownodes because failed loading originals.
Setting status of b5c4f331a12da5f699b1d1924ebad0ef2d9fffaa to FAILURE with url https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/2178/ and message: ' '
Using context: Pull Request - OpenJ9
java.io.IOException: Tried to load head FlowNodes for execution Owner[PullRequest-OpenJ9/2178:PullRequest-OpenJ9 #2178] but FlowNode was not found in storage for head id:FlowNodeId 1:93
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.initializeStorage(CpsFlowExecution.java:679)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.onLoad(CpsFlowExecution.java:716)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.getExecution(WorkflowRun.java:665)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.onLoad(WorkflowRun.java:529)
```

ashu-mehra commented:
@andrewcraik 
> Right but rather than trying to change the allocator under CS2 data structures convert the data structures to a more modern alternative (TR_ or STL) for the places that have a problem rather than changing the allocator without changing the data structure...

I still don't get it. Even if we change the CS2 data structure to TR/STL equivalent along with Region allocator, I think it will not solve the problem of failed compilations due to high memory usage as the Region, in current form, does not deallocate memory. Wouldn't it result in same behavior as this PR? Did I miss something?

andrewcraik commented:
My experience has taught me some of the CS2 data structures are wasteful in the amount of memory they grab and they use of memory internally. If you do some profiling to have a look at where the memory is going and if there are some CS2 uses that are near the top let's have a look at those and clean them up...

andrewcraik commented:
Note that CS2 is great in the right applications - there are just cases where some of the tradeoffs it makes are not idea for the use.

andrewcraik commented:
@ashu-mehra should this be kept open to be updated or closed pending rework?

ashu-mehra commented:
I guess its better to close it for now. I will come back to it later.
